### PR TITLE
ENH Uses the maximum in softmax for bfloat16

### PR DESCRIPTION
**Pull request by thomasjpfan**
Follow up to #60371

This PR uses the maximum when computing the softmax for bfloat16.

### CODE REVIEW

**Comment by cpuhrsch** [comment]
@leslie-fang-intel - Is the current behavior incorrect?

**Comment by leslie-fang-intel** [comment]
@thomasjpfan Thanks for point it out.
@cpuhrsch I think so. Sorry, I miss the `subtraction of max value` operation in the BF16 vec path. This operation exists in the bf16 tail case and fp32 vec paths.
One interesting thing, it didn't cause much accuracy difference w and w/o it. @thomasjpfan Do you have a test to reproduce the error? Can you add some test cases for it?

**Comment by cpuhrsch** [comment]
@leslie-fang-intel - shouldn't this be reproducible if the input is sufficiently small so that *only* the tail case is hit?

**Comment by leslie-fang-intel** [comment]
With printing some debugging message, I think it has hit the bf16 vec path in our [UT](https://github.com/leslie-fang-intel/pytorch/blob/e2048d641d7835db430e4030c2494aa9f7d34277/test/test_nn.py#L16404-L16407)
Similar thing has been observed with numpy w/wo using the max value:
```
>>> import numpy as np
>>> a = [1.0, 2.0, 3.0, 4.0, 1.0, 2.0, 3.0]
>>> np.exp(a) / np.sum(np.exp(a))
array([0.02364054, 0.06426166, 0.1746813 , 0.474833  , 0.02364054,
       0.06426166, 0.1746813 ])
>>> np.exp(a-np.amax(a)) / np.sum(np.exp(a-np.amax(a)))
array([0.02364054, 0.06426166, 0.1746813 , 0.474833  , 0.02364054,
       0.06426166, 0.1746813 ])
```
Referring to this [link](https://jamesmccaffrey.wordpress.com/2016/03/04/the-max-trick-when-computing-softmax/), the using of max value is for numeric stability when input of `exp()` is very large. And this case is not covered now.

**Comment by cpuhrsch** [comment]
@leslie-fang-intel - Makes sense. While fixing this could we add a test that fails deterministically due to this bug? I want to avoid accidental reverts or changes, since this seems to not be easily detected. Let's land this fix first, but as a follow up item we should have a test that verifiably prevents this.

**Review by cpuhrsch — APPROVED** [review]

**Review by leslie-fang-intel — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
> @leslie-fang-intel - Makes sense. While fixing this could we add a test that fails deterministically due to this bug? I want to avoid accidental reverts or changes, since this seems to not be easily detected. Let's land this fix first, but as a follow up item we should have a test that verifiably prevents this.

Sure, Let me figure out a case later.

**Comment by cpuhrsch** [comment]
@leslie-fang-intel - I'm going to revert https://github.com/pytorch/pytorch/pull/60371 for now since we're close to a release and want to prevent shipping bugs.

@thomasjpfan thanks for finding this and raising this PR! We'd have missed this.

**Comment by ngimel** [comment]
> ```python
> input = torch.as_tensor([[0.1, 0.2]], dtype=dtype)
> ```

This is correct behavior, because input + 100 is [100,100] in bfloat16:
```
In [10]: inp = torch.tensor([100.1, 100.2], dtype=torch.bfloat16)

In [11]: inp == 100
Out[11]: tensor([True, True])

In [12]: inp = torch.tensor([100.1, 100.2], dtype=torch.float16)

In [13]: inp == 100
Out[13]: tensor([False, False])
```
It's surprising how little precision there is in 7 mantissa bits (rule of thumb - 2-3 decimal digits)

**Comment by cpuhrsch** [comment]
@leslie-fang-intel - maybe [these tests](https://github.com/pytorch/pytorch/blob/master/test/test_nn.py) can help, in particular these lines.

```
    @skipCUDAIfRocm
    @dtypesIfCUDA(torch.half, torch.float)
    @dtypes(torch.float)
    def test_softmax_results(self, device, dtype):
        # Non-even sizes and non-zero shifts test fallback paths in vectorized kernel
        # Note: dim1 > 1024 is needed to exercise the vectorized (non-persistent) path, (16, 30576) is BERT-esque
        sizes = [(0, 10), (32, 20), (10, 0), (31, 20), (32, 21), (31, 23), (32, 1536), (31, 2048), (33, 2049), (16, 30576)]
        shifts = [(0, 0), (1, 0), (0, 1), (1, 1)]
        for fn in [F.softmax, F.log_softmax]:
            for size in sizes:
                for shift in shifts:
                    input = torch.rand(size, device=device, dtype=dtype)
                    # Note: With the largest tests we can hit upper limit of fp16 when we
                    # sum, so scale the input down to stay in a nicer range.
                    if dtype == torch.float16:
                        input = input / 100.
                    input = input[shift[0]:, shift[1]:]
                    # Note; Don't want to bprop back through slice op
                    input = input.detach().requires_grad_(True)
                    ref_input = input.clone().cpu().detach().requires_grad_(True)
                    for dim in [0, 1]:
                        ref_output = fn(ref_input, dtype=torch.float, dim=dim)
                        output = fn(input, dtype=torch.float, dim=dim)
                        grad_output = torch.rand(size, device=device, dtype=dtype)
                        grad_output = grad_output[shift[0]:, shift[1]:]
                        ref_grad_output = grad_output.clone().cpu().detach()
                        grad_input, = torch.autograd.grad(output, input, grad_outputs=(grad_output), create_graph=True)
                        ref_grad_input, = torch.autograd.grad(ref_output, ref_input,
                                                              grad_outputs=(ref_grad_output), create_graph=True)
                        grad_input.sum().backward()
                        ref_grad_input.sum().backward()

                        self.assertEqual(output, ref_output)
                        self.assertEqual(grad_input, ref_grad_input)
                        self.assertEqual(input.grad, ref_input.grad)
```

**Comment by leslie-fang-intel** [comment]
Thanks for the comments:
* I think If you are testing `output = F.softmax(input, dim=-1)` with specific `dim=-1`, it will use  https://github.com/pytorch/pytorch/blob/d2eda7f2f3d3121068aaa9fa8410d9b1fc34fe53/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L98-L103 instead of the path in PR https://github.com/pytorch/pytorch/pull/60371 which only covers `dim != -1` cases.
* As for `torch.tensor([100.1, 100.2], dtype=torch.bfloat16) == 100.0`, Please check the convert which do see some accuracy lost in this case. https://github.com/pytorch/pytorch/blob/d2eda7f2f3d3121068aaa9fa8410d9b1fc34fe53/c10/util/BFloat16-inl.h#L9-L17
* Add a test case with large input value(https://github.com/pytorch/pytorch/pull/63132). If we don't substract the max value in the calculation, It will fail deterministically due to exp(large_value) overflow.

**Comment by thomasjpfan** [comment]
Closing in flavor of #63132
